### DISCUSSION TITLE
ERROR MESSAGE: Error converting XML to ER7 after migrating from mirth 3.5.1 to mirth 4.0.1

### DISCUSSION THREAD

**Discussion by marpagno** [nextgenhealthcare/connect#5182] [Q&A] [answered]
Hi,
  I'm trying to upgrade my mirth 3.5.1 channels to mirth 4.0.1. The first issue I encountered is in a source transformer converting XML-HL7 messages to piped ER7: while all was fine in 3.5.1, in 4.0.1 I get a conversion error 

ERROR MESSAGE: Error converting XML to ER7
ca.uhn.hl7v2.HL7Exception: Namespace URI must be urn:hl7-org:v2xml

I tried re-adding the namespace at the very ending on the transformer

tmp=msg;
// .... additional  mappings 
default xml namespace = 'urn:hl7-org:v2xml';
tmp=new XML(tmp.toString());
logger.debug(tmp.toString());

[new.4.txt](https://github.com/nextgenhealthcare/connect/files/8659671/new.4.txt)

And tried also changing some knobs in the outbound parser configuration, however the only thing the seems to work a bit is disabling the strict parser in the deserialization - which produces a pseudo-HL7 all-in-one-line, preceded by a the wrapping xml ADT_A03 tag

![Immagine 2022-05-10 100318](https://user-images.githubusercontent.com/50516275/167582033-8ad57b87-0efe-4650-9999-32cf59a6d75d.png)

any other configuration I tried simply gives the parsing exception.

The channel was in production flawlessy on mirth 3.5, I'm looking for a quick workaround!

Any idea is very much appreciated!

**Comment by marpagno** [comment, marked as answer]
seems the issue can be resolved by 
1) adding at the end of the source transformer

default xml namespace = 'urn:hl7-org:v2xml';
tmp=new XML(tmp.toString());

2) adding an outbound template in the same transformer (I don't really know why it wasn't there)
MSH||||||||||||||

Now this works - still facing another issue on a channel handling MDM messages, which after applying the same tricks now states 
"ERROR MESSAGE: Error converting XML to ER7
ca.uhn.hl7v2.parser.EncodingNotSupportedException: Determine encoding for message. The following is the first 50 chars of the message for reference, although this may not be where the issue is: <HL7Message xmlns="urn:hl7-org:v2xml"><MSH><MSH.1/
	at ca.uhn.hl7v2.parser.Parser.parse(Parser.java:201)
	at com.mirth.connect.plugins.datatypes.hl7v2.ER7Serializer.fromXML(ER7Serializer.java:246)"

**Reply by marpagno** [reply]
the second issue was resolved by explicitly mapping msg to tmp:

tmp=msg;
default xml namespace = 'urn:hl7-org:v2xml';
tmp=new XML(tmp.toString();

Once again, probably bad habits but NONE OF THESE TRICKS WHERE NEEDED when the channel was running on Mirth 3.5.1

**Reply by pacmano1** [reply]
Can you post an MDM that exhibits that behavior?

**Reply by marpagno** [reply]
Yes, here it is.

[mdm.txt](https://github.com/nextgenhealthcare/connect/files/8661634/mdm.txt)

However, looking at the transformer, the code was

// eliminazione "alla bruttissima"del multipart dall'obx.5.5 - si noti che l'OBSERVATION ¿ tolta nel preprocessor
var s = msg['OBX']['OBX.5']['ED.5'].toString();
for (var i = 0; i < 9; i++) 	s = s.substr(s.indexOf("\n") + 1);
s = s.replace(/--_StringPartSeparator--/, "").replace(/\n/g, '');
msg['OBX']['OBX.5']['ED.5'] = s; // ributto sul source, se metto nell'outbound la conversione va a pallino - il che è un baco di mirth 3.5.1

//these two lines were added to fix 4.0.1 vs 3.5.1 behaviour
default xml namespace = 'urn:hl7-org:v2xml';
tmp=new XML(msg.toString());


So as you can see, there was no reference to tmp variable, when I put the two lines 
"default xml namespace = 'urn:hl7-org:v2xml';
tmp=new XML(**tmp**.toString());" 

and the reason was that, translating from (my) remark  "// ributto sul source, se metto nell'outbound la conversione va a pallino - il che è un baco di mirth 3.5.1"  that modifying the tmp gave an error on 3.5.1

ok, I know, the message with that mime multipart inside the obx.5 is a monster,
the idea of modifying the "immutable" msg was a dirty workaround taken years ago, 
relying on the implicit msg ->tmp assignment performed by mirth at the end of any empty transformer was questionable,
assigning tmp to itself without defining it is nonsense so the error (thought a bit misleading) makes sense

I think it's more interesting for others (less specific of this quick and dirty implementation I mean) the need of explicitly specifying the namespace and the outbound template (thought the latter was needed quite always, I'm surprised that worked for so many years without it)

**Reply by pacmano1** [reply]
I meant the original MDM, not the XML representation of it.

**Reply by marpagno** [reply]
Sorry, maybe I wasn't clear - the original message **is** XML, I'm translating to piped form- that is the transofmation which gave me some trouble
